src/coreclr/runtime/i386/AllocFast.S

```diff
@@ -0,0 +1,275 @@
+        // 32-bit value. We're going to tail-call to a managed helper that will throw
+        // an OOM exception that the caller of this allocator understands.
+
+        // ecx holds MethodTable pointer already
+        xor         edx, edx            // Indicate that we should throw OOM.
+        jmp         RhExceptionHandling_FailedAllocation
+LEAF_END RhNewString, _TEXT
+
+
+// Allocate one dimensional, zero based array (SZARRAY).
+//  ECX == MethodTable
+//  EDX == element count
+LEAF_ENTRY RhpNewArrayFast, _TEXT
+        NEW_ARRAY_FAST_PROLOG
+
+        // Compute overall allocation size (align(base size + (element size * elements), 4)).
+        // if the element count is <= 0x10000, no overflow is possible because the component size is
+        // <= 0xffff, and thus the product is <= 0xffff0000, and the base size for the worst case
+        // (32 dimensional MdArray) is less than 0xffff.
+        movzx       eax, word ptr [ecx + OFFSETOF__MethodTable__m_usComponentSize]
+        cmp         edx, 0x010000
+        ja          LOCAL_LABEL(RhpNewArrayFast_ArraySizeBig)
+        mul         edx
+        add         eax, [ecx + OFFSETOF__MethodTable__m_uBaseSize]
+        add         eax, 3
+LOCAL_LABEL(RhpNewArrayFast_ArrayAlignSize):
+        and         eax, -4
+
+        NEW_ARRAY_FAST
+
+LOCAL_LABEL(RhpNewArrayFast_ArraySizeBig):
+        // Compute overall allocation size (align(base size + (element size * elements), 4)).
+        // if the element count is negative, it's an overflow, otherwise it's out of memory
+        cmp         edx, 0
+        jl          LOCAL_LABEL(RhpNewArrayFast_ArraySizeOverflow)
+        mul         edx
+        jc          LOCAL_LABEL(RhpNewArrayFast_ArrayOutOfMemoryNoFrame)
+        add         eax, [ecx + OFFSETOF__MethodTable__m_uBaseSize]
+        jc          LOCAL_LABEL(RhpNewArrayFast_ArrayOutOfMemoryNoFrame)
+        add         eax, 3
+        jc          LOCAL_LABEL(RhpNewArrayFast_ArrayOutOfMemoryNoFrame)
+        jmp         LOCAL_LABEL(RhpNewArrayFast_ArrayAlignSize)
+
+LOCAL_LABEL(RhpNewArrayFast_ArrayOutOfMemoryNoFrame):
+        ESP_EPILOG_FREE 8
+
+        // ecx holds MethodTable pointer already
+        xor         edx, edx        // Indicate that we should throw OOM.
+        jmp         RhExceptionHandling_FailedAllocation
+
+LOCAL_LABEL(RhpNewArrayFast_ArraySizeOverflow):
+        ESP_EPILOG_FREE 8
+
+        // We get here if the size of the final array object can't be represented as an unsigned
+        // 32-bit value. We're going to tail-call to a managed helper that will throw
+        // an overflow exception that the caller of this allocator understands.
+
+        // ecx holds MethodTable pointer already
+        mov         edx, 1          // Indicate that we should throw OverflowException
+        jmp         RhExceptionHandling_FailedAllocation
+LEAF_END RhpNewArrayFast, _TEXT
+
+
+#ifndef FEATURE_NATIVEAOT
```

Comment:
Anything preventing enabling this optimized path for NAOT too? It should be just a few lines in ilc to enable.